rust/connlib/tunnel/src/peer.rs

```diff
@@ -335,6 +535,87 @@ pub struct ClientOnGateway {
     ipv6: Ipv6Addr,
     resources: HashMap<ResourceId, Vec<ResourceOnGateway>>,
     filters: IpNetworkTable<FilterEngine>,
+    permanent_translations: HashMap<IpAddr, (IpAddr, ResourceId, DomainName, Instant, bool)>,
+    nat_table: NatTable,
+    buffered_events: VecDeque<GatewayEvent>,
+}
+
+#[derive(Default, Debug)]
+struct NatTable {
+    table: BiMap<(Protocol, IpAddr), (Protocol, IpAddr)>,
+    last_seen: HashMap<(Protocol, IpAddr), Instant>,
+}
+
+impl NatTable {
+    fn expire(&mut self, now: Instant) {
+        let mut removed = Vec::new();
+        for (r, e) in self.last_seen.iter() {
+            if now.duration_since(*e) >= Duration::from_secs(60) {
+                self.table.remove_by_right(r);
+                removed.push(*r);
+            }
+        }
+
+        for r in removed {
+            self.last_seen.remove(&r);
+        }
+    }
+
+    fn translate(
+        &mut self,
+        outgoing_pkt: &IpPacket,
+        real_address: IpAddr,
+        now: Instant,
+    ) -> Option<(Protocol, IpAddr)> {
+        let source_protocol = outgoing_pkt.source_protocol()?;
+        if let Some(res) = self
+            .table
+            .get_by_left(&(source_protocol, outgoing_pkt.destination()))
+        {
+            if res.1 == real_address {
+                self.last_seen.insert(*res, now);
+                return Some(*res);
+            }
+        }
+
+        let mut occupied_ports = self
+            .table
+            .iter()
+            .filter(|(_, (proto, ip))| *ip == real_address && proto.same_type(&source_protocol))
+            .map(|(_, (proto, _))| proto.value())
+            .sorted_unstable();
+
+        for p in 1.. {
+            if !occupied_ports.contains(&p) {
+                let proxy_protocol = source_protocol.with_value(p);
+                self.table.insert(
+                    (source_protocol, outgoing_pkt.destination()),
+                    (proxy_protocol, real_address),
+                );
+                self.last_seen.insert((proxy_protocol, real_address), now);
+                return Some((proxy_protocol, real_address));
+            }
+        }
+
+        tracing::warn!("available nat ports exhausted");
+        None
+    }
+
+    fn revert_translation(
+        &mut self,
+        incoming_packet: &IpPacket,
+        now: Instant,
+    ) -> Option<(Protocol, IpAddr)> {
+        if let Some((destination_protocol, src)) = self.table.get_by_right(&(
+            incoming_packet.destination_protocol()?,
+            incoming_packet.source(),
+        )) {
+            self.last_seen.insert((*destination_protocol, *src), now);
+            return Some((*destination_protocol, *src));
+        }
+
+        None
```

Comment:
Add `trace` logs here for dropped and translated packets.

Reply:
packets here are not dropped, just not translated

Reply:
do we want to add a trace here for every translated packet?

Reply:
Yeah I think that would be useful, esp. for the logs in our tests. They are basically never enabled in prod but it is nice to see what happens when running `tunnel_test`.